src/sinks/signalfx/metrics.rs

```diff
@@ -0,0 +1,368 @@
+    let pairs: Vec<signalfx_proto::Dimension> = tags
+        .iter()
+        .map(|(name, value)| signalfx_proto::Dimension {
+            key: Some(name.to_string()),
+            value: Some(value.to_string()),
+        })
+        .collect();
+    pairs
+}
+
+fn encode_timestamp(timestamp: Option<DateTime<Utc>>) -> i64 {
+    match timestamp {
+        Some(ts) => ts.timestamp_millis(),
+        None => Utc::now().timestamp_millis(),
+    }
+}
+
+fn serialize_datapoints(dpu: &signalfx_proto::DataPointUploadMessage) -> Vec<u8> {
+    let mut buf = Vec::new();
+    buf.reserve(dpu.encoded_len());
+    dpu.encode(&mut buf).unwrap();
+    buf
+}
+
+fn encode_namespace(namespace: &str, name: &str) -> String {
+    if !namespace.is_empty() {
+        format!("{}.{}", namespace, name)
+    } else {
+        name.to_string()
+    }
+}
+
+fn encode_events(
+    events: Vec<Metric>,
+    interval: i64,
+    namespace: &str,
+) -> signalfx_proto::DataPointUploadMessage {
+    let series = events
+        .into_iter()
+        .filter_map(|event| {
+            let fullname = encode_namespace(namespace, &event.name);
+            let ts = encode_timestamp(event.timestamp);
+            let dimensions = match event.tags.clone().map(encode_dimensions) {
+                Some(v) => v,
+                None => vec![],
+            };
+
+            match event.kind {
+                MetricKind::Incremental => match event.value {
+                    MetricValue::Counter { value } => Some(vec![signalfx_proto::DataPoint {
+                        source: Some(String::from("vector")),
+                        metric: Some(fullname),
+                        metric_type: Some(signalfx_proto::MetricType::Counter as i32),
+                        timestamp: Some(ts),
+                        value: Some(signalfx_proto::Datum {
+                            double_value: Some(value),
+                            str_value: None,
+                            int_value: None,
+                        }),
+                        dimensions,
+                    }]),
+                    _ => None,
+                },
+                MetricKind::Absolute => match event.value {
+                    MetricValue::Gauge { value } => Some(vec![signalfx_proto::DataPoint {
+                        source: Some(String::from("vector")),
+                        metric: Some(fullname),
+                        metric_type: Some(signalfx_proto::MetricType::Gauge as i32),
+                        timestamp: Some(ts),
+                        value: Some(signalfx_proto::Datum {
+                            double_value: Some(value),
+                            str_value: None,
+                            int_value: None,
+                        }),
+                        dimensions,
+                    }]),
+                    _ => None,
+                },
+            }
+        })
+        .flatten()
+        .collect::<Vec<_>>();
```

Comment:
It looks like the inner vectors produced by this iterator are all either single elements or `None`. Do I see that right? If so, dropping the inner vector and the trailing `flatten()` seems the best route.